File: php/largest-series-product/LargestSeriesProduct.php
<?php
class Series
{
    public function __construct(private $digits) {}
    public function largestProduct($span) {
        
        if ($span < 0 || strlen($this->digits) < $span || ($this->digits && !is_numeric($this->digits))) {
            throw new \InvalidArgumentException;
        }
        if (!strlen($this->digits) || !$span) {
            return 1;
        }
        
        $max = 0;
        for ($i = 0; $i < strlen($this->digits) - $span + 1; ++$i) {
            $max = max($max, array_product(str_split(substr($this->digits, $i, $span))));
        }
        return $max;
    }
}

File: nim/largest-series-product/largest_series_product.nim
proc largestProduct*(input: string, span: int): int =
  if span notin 0..input.len:
    raise newException(ValueError, "bounds")
  for i in 0..input.len - span:
    var product = 1
    for j in 0..<span:
      let digit = ord(input[i + j]) - ord('0')
      if digit notin 0..9:
        raise newException(ValueError, "wrong series")
      product *= digit
    result = max(result, product)

